Implement new aurastate handling and start creating new MIT Spell::canCast()

Appled opened the pull request:
- Spells: Implement real aurastate handling
   - Now spells and talents with effects such as "when below X % health all damage done is increased by..." etc should work properly
- Spells: Create MIT Spell file and start implementing new canCast
- Spells: Move aurastates and auraeffects into their own header files
- Players: Verify player flags for 3.3.5a

Updated version of https://github.com/AscEmu/AscEmu/pull/536

Sorry it took longer than I thought. I had to do tons of research for it.
Thanks to @Zyres for final player flag verifications.

Appled commented:
I had to add version checks in one function and the function now only returns false on cata. https://github.com/Appled/AscEmu/blob/9168eb6983f714d60f68dc8fa302928bb3ecc4ee/src/world/Spell/SpellInfo.cpp#L485
I'm too tired to fix it properly now, I'll try to fix it in next PR.

schnek commented:
It is possible in the future to separate spells by class and branch
Spell / Warlock /
------------------  Affliction.cpp
------------------  Demonology.cpp
------------------  Destruction.cpp
idea* :octocat:

Evairfairy commented on src/world/Spell/Spell.Legacy.cpp:
```diff
@@ -926,17 +926,16 @@ uint8 Spell::prepare(SpellCastTargets* targets)
     if (cancastresult != SPELL_CANCAST_OK)
     {
-        SendCastResult(cancastresult);
+        // Triggered spells also need to go through cancast check but they do not pop a error message
```
Why? What incorrect behaviour is caused by skipping this check?

Appled replied:
Nothing is skipped? Triggered spells also need to be checked in canCast() (for range, aurastate etc) but they just do not send error message to client. Before this change triggered spells weren't even checked in canCast(). See https://github.com/AscEmu/AscEmu/pull/572/files#diff-a5339c3ab8259ee5d6d4a3a957d11f80L929

Evairfairy replied:
@Appled That's what I'm referring to.

Previously, these checks were skipped for triggered spells, however now you're wanting to apply them.

Are you certain that's the correct behaviour?

Appled replied:
@Evairfairy Oh I've managed to misunderstand your question then. There are certain things which must be checked in canCast (which are not yet implemented), such as death state for triggered spells which didn't proc from an aura, target checks, power checks etc. However legacy canCast is not called for triggered spells, see here https://github.com/Appled/AscEmu/blob/9168eb6983f714d60f68dc8fa302928bb3ecc4ee/src/world/Spell/Spell.cpp#L95

Evairfairy commented on src/world/GameCata/Handlers/CharacterHandler.cpp:
```diff
@@ -115,15 +115,15 @@ void WorldSession::CharacterEnumProc(QueryResult* result)
+                char_flags |= PLAYER_FLAG_PREVENT_MELEE_SPELLS;
             if (charEnum.deathState != 0)
-                char_flags |= PLAYER_FLAG_IS_DEAD;
+                char_flags |= PLAYER_FLAG_PLAYED_5_HOURS;
```
If these flags are incorrect, please add a comment explaining your change.

Appled replied:
I tested these on wotlk and I forgot to include comments to other expansions. The flag value isn't changed, only the flag is renamed properly. But like you might've guessed, character screen does not use player flags.
**However this change has no impact ingame.** Correcting these character flags to their proper names is next on my to-do list. Although I have yet to verify and test them all.
I will include the comments from wotlk though.

Evairfairy commented on src/world/GameClassic/Handlers/CharacterHandler.cpp:
```diff
@@ -121,15 +121,15 @@ void WorldSession::CharacterEnumProc(QueryResult* result)
+                char_flags |= PLAYER_FLAG_PREVENT_MELEE_SPELLS;
             if (charEnum.deathState != 0)
-                char_flags |= PLAYER_FLAG_IS_DEAD;
+                char_flags |= PLAYER_FLAG_PLAYED_5_HOURS;
```
If these flags are incorrect, please add a comment explaining your change.

Appled replied:
See https://github.com/AscEmu/AscEmu/pull/572/files#r174584860

Evairfairy commented on src/world/GameTBC/Handlers/CharacterHandler.cpp:
```diff
@@ -122,15 +122,15 @@ void WorldSession::CharacterEnumProc(QueryResult* result)
+                char_flags |= PLAYER_FLAG_PREVENT_MELEE_SPELLS;
             if (charEnum.deathState != 0)
-                char_flags |= PLAYER_FLAG_IS_DEAD;
+                char_flags |= PLAYER_FLAG_PLAYED_5_HOURS;
```
If these flags are incorrect, please add a comment explaining your change.

Appled replied:
See https://github.com/AscEmu/AscEmu/pull/572/files#r174584860

Evairfairy commented on src/world/GameWotLK/Handlers/CharacterHandler.cpp:
```diff
@@ -123,15 +123,15 @@ void WorldSession::CharacterEnumProc(QueryResult* result)
+                char_flags |= PLAYER_FLAG_PREVENT_MELEE_SPELLS; // Apparently character screen does not use player flags, debug these (this flag shows player is banned)
             if (charEnum.deathState != 0)
-                char_flags |= PLAYER_FLAG_IS_DEAD;
+                char_flags |= PLAYER_FLAG_PLAYED_5_HOURS; // Apparently character screen does not use player flags, debug these (this flag shows player is dead)
```
Please do not use inline comments. Place the comment on the line above.

Evairfairy commented on src/world/Spell/Customization/HackFixes.cpp:
```diff
@@ -692,14 +692,6 @@ void ApplyNormalFixes()
-            sp->setCasterAuraState(1 << (sp->getCasterAuraState() - 1));
-
-        if (sp->getTargetAuraState() > 1)
-            sp->setTargetAuraState(1 << (sp->getTargetAuraState() - 1));
```
Please explain this change.

Appled replied:
Well, Spell.dbc stores aurastate data in decimal since the aurastate fields can each have only one aurastate. The removed change here converted them to bitmask, which is not needed. A single spell cannot require more than one aurastate from caster or target.

Evairfairy commented on src/world/Spell/Spell.Legacy.cpp:
```diff
@@ -988,10 +987,15 @@ uint8 Spell::prepare(SpellCastTargets* targets)
+
         // aura state removal
-        if (GetSpellInfo() && GetSpellInfo()->getCasterAuraState() != AURASTATE_NONE && GetSpellInfo()->getCasterAuraState() != AURASTATE_FLAG_JUDGEMENT)
+        /*if (GetSpellInfo() && GetSpellInfo()->getCasterAuraState() != AURASTATE_NONE && GetSpellInfo()->getCasterAuraState() != AURASTATE_FLAG_JUDGEMENT)
```
Please do not comment out blocks of code - if the code is not supposed to be there, remove it.

Evairfairy commented on src/world/Spell/Spell.cpp:
```diff
@@ -0,0 +1,96 @@
+
+    if (u_caster != nullptr)
+    {
+        bool requireCombat = true;
```
Please use auto instead of explicitly specifying the type.

Evairfairy commented on src/world/Spell/Spell.cpp:
```diff
@@ -0,0 +1,96 @@
+        bool requireCombat = true;
+        if (u_caster->hasAuraWithAuraEffect(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+        {
+            for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
```
Why is this a u32? Please use auto for implicit typing, or size_t to index arrays.

Evairfairy commented on src/world/Spell/Spell.cpp:
```diff
@@ -0,0 +1,96 @@
+    {
+        // You can't cast other spells if you have the player flag preventing cast
+        if (p_caster->hasPlayerFlags(PLAYER_FLAG_PREVENT_SPELL_CAST))
+        {
```
For this condition (and all similar single line returns) please omit the braces.

Appled replied:
I think it was mentioned in the wiki under codestyle to not omit braces for single line if-statements 😃 I can be mistaken though.

Evairfairy replied:
@Appled It's entirely possible, the code style hasn't been updated in a while.

Evairfairy commented on src/world/Spell/SpellAuras.cpp:
```diff
@@ -955,15 +955,12 @@ void Aura::Remove()
+        m_target->removeAuraState(AURASTATE_FLAG_BLEED);
     if (m_spellInfo->custom_BGR_one_buff_on_target & SPELL_TYPE_SEAL && !--m_target->asc_seal)
-        flag |= AURASTATE_FLAG_JUDGEMENT;
-    if (flag != 0)
```
Please explain the removal of this block.

Appled replied:
Before this change aurastates were handled in bitmasks. The block is just made "decimal-friendly".

Evairfairy commented on src/world/Spell/SpellAuras.cpp:
```diff
@@ -1805,22 +1802,22 @@ void Aura::SpellAuraPeriodicDamage(bool apply)
         {
-            m_target->SetFlag(UNIT_FIELD_AURASTATE, AURASTATE_FLAG_IMMOLATE);
+            m_target->addAuraState(AURASTATE_FLAG_CONFLAGRATE);
         }
```
Please explain these flag changes:

- from immolate to conflagrate
- from poison to envenom
- from rejuvenate to swiftmend

Appled replied:
I will add definitions for them in AuraStates.h

Evairfairy commented on src/world/Spell/SpellInfo.cpp:
```diff
@@ -480,6 +480,35 @@ int SpellInfo::firstBeneficialEffect() const
+    // TODO: Verify for cata
+#if VERSION_STRING != Cata
+    if (spellInfo == nullptr)
+    {
```
For this and all similar single line returns, please omit the braces.

Evairfairy commented on src/world/Spell/SpellInfo.cpp:
```diff
@@ -480,6 +480,35 @@ int SpellInfo::firstBeneficialEffect() const
+        return false;
+    }
+
+    for (uint8_t i = 0; i < MAX_SPELL_EFFECTS; ++i)
```
Please use auto for implicit typing for size_t for indexing arrays.

Evairfairy commented on src/world/Spell/Spell_ClassScripts.cpp:
```diff
@@ -81,9 +81,9 @@ class FireNova : public Spell
+    SpellCastResult canCast(bool tolerate)
     {
-        uint8 result = Spell::CanCast(tolerate);
+        SpellCastResult result = Spell::canCast(tolerate);
```
Is explicit typing required here? If not, please use implicit typing.

Appled replied:
Not required, but canCast returns value in SpellCastResult and not uint8.

Evairfairy replied:
@Appled but what I mean is you should only be specifying the type if auto is evaluating to the wrong type, and I don't see why it would be here.

Evairfairy commented on src/world/Units/Players/Player.Legacy.cpp:
```diff
@@ -14000,10 +13986,10 @@ void Player::DealDamage(Unit* pVictim, uint32 damage, uint32 /*targetEvent*/, ui
 
                 if (!sEventMgr.HasEvent(this, EVENT_LASTKILLWITHHONOR_FLAG_EXPIRE))
-                    sEventMgr.AddEvent(static_cast<Unit*>(this), &Unit::EventAurastateExpire, static_cast<uint32>(AURASTATE_FLAG_LASTKILLWITHHONOR), EVENT_LASTKILLWITHHONOR_FLAG_EXPIRE, 20000, 1, 0);
+                    sEventMgr.AddEvent(static_cast<Unit*>(this), &Unit::removeAuraState, AURASTATE_FLAG_LASTKILLWITHHONOR, EVENT_LASTKILLWITHHONOR_FLAG_EXPIRE, 20000, 1, 0);
```
Please explain why you removed the cast here.

This also applies to every other place in which you have removed the cast in the AddEvent call.

Appled replied:
EventAurastateExpire had a parameter in uint32 just because the aurastates were handled in bitmask. However this is not the case anymore.

Evairfairy commented on src/world/Units/Players/PlayerDefines.hpp:
```diff
@@ -309,39 +309,42 @@ enum Standing
-    PLAYER_FLAG_UNK13               = 0x40000000,
-    PLAYER_FLAG_UNK14               = 0x80000000
+    PLAYER_FLAG_NONE                    = 0x00000000,
+    PLAYER_FLAG_PARTY_LEADER            = 0x00000001, //(TODO: implement for all versions) Informs players outside of your group who is your group leader
```
Please add a space between the // and the comment itself. The same applies to lines 330 and 340.

Evairfairy commented on src/world/Units/Unit.Legacy.cpp:
```diff
@@ -9066,15 +9058,12 @@ void Unit::AddAura(Aura* aur)
     if (aur->GetSpellInfo()->custom_BGR_one_buff_on_target & SPELL_TYPE_SEAL && !asc_seal++)
-        flag |= AURASTATE_FLAG_JUDGEMENT;
-
-    SetFlag(UNIT_FIELD_AURASTATE, flag);
```
Why have you removed this call?

Zyres replied:
because addAuraState adds the flag directly.

Appled replied:
Same as https://github.com/AscEmu/AscEmu/pull/572/files#r174589664

Evairfairy commented on src/world/Units/Unit.Legacy.cpp:
```diff
@@ -12237,21 +12223,20 @@ void Unit::EventHealthChangeSinceLastUpdate()
 #if VERSION_STRING >= WotLK
-    if (pct < 75)
-        RemoveFlag(UNIT_FIELD_AURASTATE, AURASTATE_FLAG_HEALTH75);
+    if (pct > 75)
```
Is this change intentional?

Appled replied:
Yes, the health75 state was added in wotlk and is actually above 75% health and not below like other health related aurastates :)

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
 
+bool Unit::hasAuraWithAuraEffect(AuraEffect type) const
+{
+    for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
```
Please use auto for implicit typing, or size_t to index arrays.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+{
+    if (caster != nullptr && spellInfo != nullptr && caster->hasAuraWithAuraEffect(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+    {
+        for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
```
Please use auto for implicit typing, or size_t to index arrays.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+{
+    if (!(unitData()->aura_state & (1 << (state - 1))))
+    {
+        write(unitData()->aura_state, unitData()->aura_state | uint32_t(1 << (state - 1)));
```
Please isolate the write call into its own function and then call that, instead of mixing it with other logic.

Zyres replied:
We will run into some naming issues here... addAuraState should be used to isolate the write function (unitData()).
This function does not simply add the aurastates.... it does also controle (checks) if the spell can be casted....
I would suggest to rename this function to addAuraStateAndCast so the write(unitData()) part can be isolated to addAuraState.

Evairfairy replied:
@Appled I agree with Zyres' suggestion, assuming you're able to perform method rename refactoring.

If not, let us know and one of us will do it.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+        if (IsPlayer())
+        {
+            // Activate passive spells which require this aurastate
+            SpellSet const playerSpellMap = static_cast<Player*>(this)->mSpells;
```
Please use implicit typing.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+        {
+            // Activate passive spells which require this aurastate
+            SpellSet const playerSpellMap = static_cast<Player*>(this)->mSpells;
+            for (const uint32_t spellId : playerSpellMap)
```
Please use implicit typing.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+        write(unitData()->aura_state, unitData()->aura_state & ~(1 << (state - 1)));
+        // Remove self-applied passive auras requiring this aurastate
+        // Skip removing enrage effects
+        for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
```
Please use auto for implicit typing, or size_t to index arrays.

Evairfairy commented on src/world/Units/Unit.cpp:
```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
+{
+    if (unitData()->aura_state & (1 << (state - 1)))
+    {
+        write(unitData()->aura_state, unitData()->aura_state & ~(1 << (state - 1)));
```
Please isolate the write call into its own function and then call that, instead of mixing it with other logic.

Zyres replied:
Same as mentioned a few lines above.
This function does not only write the unit data, it also removes locale saved Unit auras.
I would suggest to rename this function to removeAuraStateAndAuras and isolate the write(unitData()) part to removeAuraState function.

Appled commented:
@Evairfairy @Zyres I have added your requested changes into the PR. Please see if I've forgotten anything.
I've also isolated as requested the write function into its own function and I've renamed 'addAuraState' to 'addAuraStateAndAuras' and 'removeAuraState' to 'removeAuraStateAndAuras'.

Evairfairy commented:
Approved pending successful build.
